src/lavinmq/queue/stream_queue.cr

```diff
@@ -0,0 +1,187 @@
+    # save message id / segment position
+    def publish(msg : Message) : Bool
+      return false if @state.closed?
+      reject_on_overflow(msg)
+      @msg_store_lock.synchronize do
+        msg = add_offset_header(msg, message_count) # save last_index in RAM and update and set it as offset? #how to set first offset on startup/new queue?
+        @msg_store.push(msg)
+        @publish_count += 1
+      end
+      drop_overflow unless immediate_delivery?
+      true
+    rescue ex : MessageStore::Error
+      @log.error(exception: ex) { "Queue closed due to error" }
+      close
+      raise ex
+    end
+
+    def basic_get(no_ack, force = false, & : Envelope -> Nil) : Bool
+      return false if !@state.running? && (@state.paused? && !force)
+      @last_get_time = RoughTime.monotonic
+      @queue_expiration_ttl_change.try_send? nil
+      @get_count += 1
+      get(true) do |env|
+        yield env
+      end
+    end
+
+    # If nil is returned it means that the delivery limit is reached
+    def consume_get(no_ack, & : Envelope -> Nil) : Bool
+      get(true) do |env|
+        yield env
+        env.redelivered ? (@redeliver_count += 1) : (@deliver_count += 1)
+      end
+    end
+
+    # yield the next message in the ready queue
+    # returns true if a message was deliviered, false otherwise
+    # if we encouncer an unrecoverable ReadError, close queue
+    private def get(no_ack, & : Envelope -> Nil) : Bool
+      raise ClosedError.new if @closed
+      loop do # retry if msg expired or deliver limit hit
+        env = @msg_store_lock.synchronize { @msg_store.shift? } || break
+        if has_expired?(env.message) # guarantee to not deliver expired messages
+          expire_msg(env, :expired)
+          next
+        end
+        sp = env.segment_position
+        if no_ack
+          begin
+            yield env # deliver the message
+          rescue ex   # requeue failed delivery
+            @msg_store_lock.synchronize { @msg_store.requeue(sp) }
+            raise ex
+          end
+          delete_message(sp)
+        else
+          mark_unacked(sp) do
+            yield env # deliver the message
+          end
+        end
+        return true
+      end
+      false
+    rescue ex : MessageStore::Error
+      @log.error(exception: ex) { "Queue closed due to error" }
+      close
+      raise ClosedError.new(cause: ex)
+    end
+
+    # handle offset
+    def add_consumer(consumer : Client::Channel::Consumer)
+      return if @closed
+      @last_get_time = RoughTime.monotonic
+      @consumers_lock.synchronize do
+        was_empty = @consumers.empty?
+        @consumers << consumer
+        notify_consumers_empty(false) if was_empty
+      end
+      @exclusive_consumer = true if consumer.exclusive
+      @has_priority_consumers = true unless consumer.priority.zero?
+      @log.debug { "Adding consumer (now #{@consumers.size})" }
+      notify_observers(:add_consumer, consumer)
+    end
+
+    private def handle_arguments
```

Comment:
what's different here from the normal Queue implementation? doesnt it need to be overriden?